google/cloud/storage/async/client.h

```diff
@@ -561,6 +562,130 @@ class AsyncClient {
          /*.options=*/std::move(options)});
   }
 
+  /**
+   * Creates an `AsyncRewriter` to copy the source object.
+   *
+   * Applications use this function to reliably copy objects across [location
+   * boundaries](https://cloud.google.com/storage/docs/locations), and to
+   * rewrite objects with different encryption keys. The operation returns a
+   * `ObjectRewriter`, which the application can use to initiate the copy and to
+   * iterate if the copy requires more than one call to complete.
+   *
+   * @note Application developers should be aware that rewriting large objects
+   *     may take many calls to `AsyncRewriter::Iterate()`. Simple experiments
+   *     rewriting an object within the same bucket or rewriting small objects
+   *     often complete with a single call. For more information, see the
+   *     [Object: rewrite] documentation.
+   *
+   * @param source_bucket_name the name of the bucket containing the source
+   *     object.
+   * @param source_object_name the name of the source object.
+   * @param destination_bucket_name where the destination object will be
+   *     located.
+   * @param destination_object_name what to name the destination object.
+   * @param args a list of optional query parameters and/or request headers.
+   *     Valid types for this operation include `DestinationKmsKeyName`,
+   *      `DestinationPredefinedAcl`, `EncryptionKey`, `IfGenerationMatch`,
+   *      `IfGenerationNotMatch`, `IfMetagenerationMatch`,
+   *      `IfSourceGenerationMatch`, `IfSourceGenerationNotMatch`,
+   *      `IfSourceMetagenerationMatch`, `IfSourceMetagenerationNotMatch`,
+   *      `MaxBytesRewrittenPerCall`, `Projection`, `SourceEncryptionKey`,
+   *      `SourceGeneration`, `UserProject`, and `WithObjectMetadata`.
+   *
+   * @par Idempotency
+   * This operation is purely local, and always succeeds.  The `Iterate()` calls
+   * may call might fail and are only idempotent if restricted by
+   * pre-conditions. It this call the most relevant pre-condition is,
+   * `IfGenerationMatch`.
+   *
+   * @par Example
+   * @snippet storage_async_samples.cc rewrite-object
+   *
+   * @see
+   * [Object: rewrite]:
+   * https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite
+   */
+  template <typename... Args>
+  std::pair<AsyncRewriter, AsyncToken> StartRewrite(
+      std::string source_bucket_name, std::string source_object_name,
+      std::string destination_bucket_name, std::string destination_object_name,
+      Args&&... args) {
```

Comment:
should the implementation just call `ResumeRewrite` with an empty string for the `rewrite_token`?

Reply:
Done.